### ISSUE TITLE
Prefer if over match simple sequences and mappings

### ISSUE THREAD

**Issue by ZhdanovAM72** [wemake-services/wemake-python-styleguide#3527]
# Rule request

## Thesis
We should support simplifying `match` statements that use simple sequence and mapping patterns:

Example:
```python
match data:
    case [1, 2]:
        handle_pair()
    case _:
        ignore()
```
Can be rewritten as:
```python
if data == [1, 2]:
    handle_pair()
else:
    ignore()
```

## Reasoning
Using match for exact structural comparisons of simple literals is unnecessarily verbose. While match excels at deconstruction, using it to check for an exact list or dict value is better expressed with a direct equality comparison (`==`), which is more readable and performant.

These patterns are common in configuration checks, API responses, and state validation.

When should this be allowed?

Only when:
- There are exactly two cases (with `case _:`)
- The first pattern is a simple sequence like `[1, "a"]` or a simple mapping like `{"x": 1}`
- All elements/keys/values are simple (literals, constants, names)
- No starred patterns (`*rest`) or variable bindings (`case [x, y]:`)
- Keys in mappings are constants (not variables)

Not allowed if:
- Uses deconstruction (e.g. `case [x, y]:`)
- Contains `*args` or *`*kwargs`
- Has guards (`if` ...)

**Comment by ZhdanovAM72** [comment]
Can i take this?

**Comment by sobolevn** [comment]
@antriantsek Hi, @antriantsek! Thanks for your interest in this project.
However, there are several things I want to highlight:
1. Your PR https://github.com/wemake-services/wemake-python-styleguide/pull/3524 does not work, it is very far from the solution. If you try to run this code - it probably even won't. If you are using AI for that - please, don't :) Since there are quite a lot of things to fix in this PR, I didn't have time yet to fully review it and guide you.
2. Pinging maintainers on unrelated issues is not very polite, please don't do that :)

**Comment by antriantsek** [comment]
im sorry about that... im in the middle of a school project thats why i disturb you :)

**Comment by sobolevn** [comment]
No worries, I will happy to help, but no *right now* :)
In the meantime you can try to fix the CI yourself :heart

**Comment by Khabib73** [comment]
I'll try to do it today

**Comment by Khabib73** [comment]
Do I understand correctly that this rule should extend `SimplifiableMatchViolation` or it's new rule?

**Comment by sobolevn** [comment]
We can reuse the same one, sure :)

**Comment by Khabib73** [comment]
Hi everyone, I'm almost done!

**Comment by Khabib73** [comment]
@sobolevn 

I'm a little confused. I’d like to clarify the new rule for simplifying match statements. The rule explicitly states that simplification to if is not allowed when the first case contains a guard.

> Not allowed if:
> 
> Uses deconstruction (e.g. case [x, y]:)
> Contains *args or **kwargs
> Has guards (if ...)

So, both of these examples must NOT be simplified:

```python
match subject:
    case [1] if x > 0:
        pass
    case _:
        pass
```

and

```python
match subject:
    case x if x > 0:
        pass
    case _:
        pass
```
However, our current test suite includes a test named `simplifiable_guard_match` that expects a simplification warning for case `x if x > 0`. This contradicts the new rule.

Proposal:

- Delete or rewrite that test to expect no warning (i.e., the linter should ignore such match statements).

- Keep the logic: if we see a guard, we immediately stop and do not simplify, no matter what else is inside.

If anyone disagrees, let’s discuss
